[Nav Drawer] Move user profile and sign out link into popup #2397

NudurupatiSurya opened the pull request:
Fixes #2397 

- Modified code to show a signout popup when the user clicks on the profile photo in the NavDrawer. [issue: https://github.com/google/ground-android/issues/2397]
- This time extracted user details using `AuthenticationManager`

Updated Demo:


https://github.com/google/ground-android/assets/53263580/d6c947ab-a6d7-4c87-9610-7f7dd31c4124

gino-m commented:
Looks good! Please resolve conflicts and mark as "ready for review" once ready!

shobhitagarwal1612 commented on ground/src/debug/local/google-services.json:
Can you please revert this file?

NudurupatiSurya replied:
Sure!

NudurupatiSurya replied:
@shobhitagarwal1612, reverted the `google-services.json` back. Thank you for the feedback 😊

shobhitagarwal1612 commented:
Thanks for working on this @NudurupatiSurya. Excited to get this merged too! 🎉 🎉 🎉 

After comparing your screen recording with the [mocks](https://github.com/google/ground-android/issues/2397#issue-2194442005), I can see that there are a couple of more UI changes needed:

1. We need to show the app logo and name in the header and only show the profile photo at the right
2. Styling of this new dialog is not the same as the original signout confirmation dialog. So, could you please create a new compose dialog instead of reusing the existing one?
3. As per the mocks, the dialog appears to be anchored to the profile photo. Can you please check if that is feasible?

shobhitagarwal1612 reviewed (changes requested):
Added comments above.

gino-m commented on ground/src/main/java/com/google/android/ground/ui/home/SignOutConfirmationDialog.kt:
```diff
@@ -15,41 +15,87 @@
 
-  fun dismissDialog() {
-    openDialog.value = false
+  fun dismissSignoutDialog() {
```
This method name would then remain unchanged.

gino-m commented on ground/src/main/java/com/google/android/ground/ui/home/SignOutConfirmationDialog.kt:
```diff
@@ -15,41 +15,87 @@
 
 @Composable
-fun SignOutConfirmationDialog(openDialog: MutableState<Boolean>, signOutCallback: () -> Unit) {
+fun SignOutConfirmationDialog(
```
Sorry just noticing this now; this dialog class now represents two dialogs, not just the Sign Out confirmation (see [Single Responsibility Principle](https://en.wikipedia.org/wiki/Single_responsibility_principle)). Can you instead create a separate class called "UserDetailsDialog" and move the new dialog there?

NudurupatiSurya replied:
Sure!

gino-m reviewed (changes requested):
Apologies for late feedback.. one small refactor, otherwise LG!

shobhitagarwal1612 commented on ground/src/main/java/com/google/android/ground/ui/home/UserDetailsDialog.kt:
```diff
@@ -0,0 +1,55 @@
+    dismissButton = {
+      OutlinedButton(onClick = { dismissDialog() }) {
+        Text(
+          text = context.getString(R.string.cancel),
```
We can get the string without using `context`. Something like this: `stringResource(R.string.cancel)`

Same comment for the rest of the class.

shobhitagarwal1612 commented on ground/src/main/java/com/google/android/ground/ui/home/UserDetailsDialog.kt:
```diff
@@ -0,0 +1,55 @@
+import com.google.android.material.color.MaterialColors
+
+@Composable
+fun UserDetailsDialog(context: Context, homeScreenViewModel: HomeScreenViewModel) {
```
Instead of passing the view model, can we directly pass the user object?

Same for visibility, we can pass a `mutableState` (see SignOutConfirmationDialog for reference) and change it to false directly.

shobhitagarwal1612 commented on ground/src/main/java/com/google/android/ground/ui/home/UserDetailsDialog.kt:
```diff
@@ -0,0 +1,55 @@
+      }
+    },
+    confirmButton = {
+      TextButton(onClick = { homeScreenViewModel.showSignOutConfirmationDialog() }) {
```
Same as above. Let's not pass the view model as that would tightly couple this UI class with the view model.

gino-m replied:
+1. Applying the Principle of Least Privilege reduces changes for scope creep, the enemy of Separate of Concerns and Single Responsibility.

shobhitagarwal1612 commented on ground/src/main/java/com/google/android/ground/ui/home/UserDetailsDialog.kt:
```diff
@@ -0,0 +1,55 @@
+    confirmButton = {
+      TextButton(onClick = { homeScreenViewModel.showSignOutConfirmationDialog() }) {
+        Text(
+          text = context.getString(R.string.sign_out),
```
Same as above. We can get the string without using conext.

shobhitagarwal1612 commented on ground/src/main/java/com/google/android/ground/ui/home/UserDetailsDialog.kt:
```diff
@@ -0,0 +1,55 @@
+      TextButton(onClick = { homeScreenViewModel.showSignOutConfirmationDialog() }) {
+        Text(
+          text = context.getString(R.string.sign_out),
+          color = Color(MaterialColors.getColor(context, R.attr.colorError, "")),
```
IIRC, we don't need error color for the user details dialog. See the sample screenshot [here](https://github.com/google/ground-android/issues/2397#issue-2194442005)

NudurupatiSurya replied:
I have made this change, but I noticed that the order of the negative and positive buttons in our dialog differs from the [sample screenshot](https://github.com/google/ground-android/issues/2397). 

This is because the AlertDialog Composable has predefined positions for its `dismissButton` and `confirmButton`, and currently, there is no property to rearrange these buttons' order.

One workaround could be to assign the "Sign Out" text to the `dismissButton` and use the `confirmButton` for the "Cancel/Close" text. However, I'm concerned this might impact code readability. Please let me know your thoughts.

CC: @gino-m & @shobhitagarwal1612

gino-m replied:
> One workaround could be to assign the "Sign Out" text to the dismissButton and use the confirmButton for the "Cancel/Close" text. However, I'm concerned this might impact code readability. Please let me know your thoughts.

This would be an acceptable solution. Could you please add a comment above `dismissButton =` and `confirmButton =` clarifying the purpose of each, for ex. "Sign out button" and "Close button"

NudurupatiSurya replied:
Sure I can do that 👍

NudurupatiSurya commented:
> After comparing your screen recording with the [mocks](https://github.com/google/ground-android/issues/2397#issue-2194442005), I can see that there are a couple of more UI changes needed:
> 
> 1. We need to show the app logo and name in the header and only show the profile photo at the right
> 2. Styling of this new dialog is not the same as the original signout confirmation dialog. So, could you please create a new compose dialog instead of reusing the existing one?
> 3. As per the mocks, the dialog appears to be anchored to the profile photo. Can you please check if that is feasible?

Hello @shobhitagarwal1612,

As you suggested, I have updated the navigation header to display the app logo and name, with the profile photo positioned on the right.

This is how it looks now:

<img src="https://github.com/google/ground-android/assets/53263580/d5c1ff1d-d675-4091-bd4b-4772d13c0eb5" width="300">

However, I'm unsure about the style of the text for "Ground" and the spacing between the profile photo and the edge of the header. Could you please provide details if available?

I am also working on anchoring the user dialog to the navigation header and will convert the PR to ready for review once this is completed. Also, please let me know if you find anything that is missing.

Thank you,
Surya

CC: @gino-m

gino-m commented:
Thanks you, @NudurupatiSurya!

> However, I'm unsure about the style of the text for "Ground" and the spacing between the profile photo and the edge of the header. Could you please provide details if available?

@rawbzz, could you please advise?

rawbzz commented:
@NudurupatiSurya Ground title should be Google san, regular, 18 pts. and the distance between the avatar and the edge should be 24 px. ty!

NudurupatiSurya commented:
> @NudurupatiSurya Ground title should be Google san, regular, 18 pts. and the distance between the avatar and the edge should be 24 px. ty!

Thank you, @rawbzz! Could also you please clarify how much the dialog is anchored towards the navigation bar header? See this attached [screenshot](https://private-user-images.githubusercontent.com/228050/313995255-803ee6d8-4015-4f91-8b37-2efd3ff2f4e3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ0NDg1NTUsIm5iZiI6MTcxNDQ0ODI1NSwicGF0aCI6Ii8yMjgwNTAvMzEzOTk1MjU1LTgwM2VlNmQ4LTQwMTUtNGY5MS04YjM3LTJlZmQzZmYyZjRlMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQzMFQwMzM3MzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01Yzc1ZDUzYTU4ZTYzYjljZTU5MTgyMWFlOTMwYThlZmQwOTZmNzU3OTJmODNlYWZmOGYwODRhZjU0YWYyZTViJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.PoGfdvjNv7PMjkWyirzZGeloUGpHeMvF1aSbyCwdyqU) for reference.

gino-m commented on ground/src/main/res/layout/nav_drawer_header.xml:
```diff
@@ -62,17 +62,18 @@
           android:layout_height="wrap_content"
           android:layout_width="wrap_content"
-          android:textSize="22sp"
+          android:textSize="18pt"
```
I think we should still be using `sp` here:
https://developer.android.com/training/multiscreen/screendensities

gino-m commented on ground/src/main/res/layout/nav_drawer_header.xml:
```diff
@@ -62,17 +62,18 @@
         style="@style/NavDrawerHeader.Image"
         app:imageUrl="@{user.photoUrl}"
-        android:layout_marginEnd="4dp"
+        android:layout_marginEnd="24px"
```
And `dp` here:
https://developer.android.com/training/multiscreen/screendensities

NudurupatiSurya commented:
Hello @gino-m,

Thank you for your feedback! I've updated the font size from `pt` to `sp` and changed marginEnd from `px` to `dp` as suggested.

Sorry for the delay in my response, I was tied up with finals week at my college.

I've moved the PR from draft to ready for review. Please have a look and let me know if further adjustments are needed.

Looking forward to potentially merging my first PR 😊

Best regards,
Surya

CC: @shobhitagarwal1612

shobhitagarwal1612 commented:
Can you please update the screen recording in the PR description as well?

NudurupatiSurya commented:
> Can you please update the screen recording in the PR description as well?

Updated the description

shobhitagarwal1612 commented:
/gcbrun

shobhitagarwal1612 commented on ground/src/main/java/com/google/android/ground/ui/home/HomeScreenFragment.kt:
```diff
@@ -99,6 +100,9 @@ class HomeScreenFragment :
     navHeader.findViewById<TextView>(R.id.switch_survey_button).setOnClickListener {
       homeScreenViewModel.showSurveySelector()
     }
+    navHeader.findViewById<ShapeableImageView>(R.id.ground_logo).setOnClickListener {
+      showSignOutConfirmationDialogs()
+    }
```
The click for signout confirmation dialog should be bound to the profile image and not app logo.

shobhitagarwal1612 commented on ground/src/main/java/com/google/android/ground/ui/home/HomeScreenFragment.kt:
```diff
@@ -143,26 +147,44 @@ class HomeScreenFragment :
+        showUserDetailsDialog.value = true
+        showSignOutDialog.value = false
+
+        val user = homeScreenViewModel.userDetails.value!!
```
This fragment already injects `UserRepository`. So, we don't need to fetch the details from view model. Instead, we can do something like `userRepository.getAuthenticatedUser()` directly here.

NudurupatiSurya replied:
Thank you for the suggestion @shobhitagarwal1612! I have changed it now.

shobhitagarwal1612 commented on ground/src/main/java/com/google/android/ground/ui/home/HomeScreenViewModel.kt:
```diff
@@ -43,11 +45,19 @@ internal constructor(
   private val offlineAreaRepository: OfflineAreaRepository,
   private val submissionRepository: SubmissionRepository,
   private val surveyRepository: SurveyRepository,
+  private val authenticationManager: AuthenticationManager,
 ) : AbstractViewModel() {
 
   private val _openDrawerRequests: MutableSharedFlow<Unit> = MutableSharedFlow()
   val openDrawerRequestsFlow: SharedFlow<Unit> = _openDrawerRequests.asSharedFlow()
 
+  private val _userDetails = MutableLiveData<User>()
+  val userDetails: LiveData<User> = _userDetails
+
+  init {
+    viewModelScope.launch { _userDetails.value = authenticationManager.getAuthenticatedUser() }
+  }
+
```
As mentioned above, this is no longer needed.

shobhitagarwal1612 commented:
Also pasting the output of GCB here:

```
* What went wrong:
Execution failed for task ':workspace:ground:ktfmtCheckMain'.
> [ktfmt] Found 3 files that are not properly formatted:
  src/main/java/com/google/android/ground/ui/home/SignOutConfirmationDialog.kt
  src/main/java/com/google/android/ground/ui/home/HomeScreenFragment.kt
  src/main/java/com/google/android/ground/ui/home/UserDetailsDialog.kt
```

You can fix this locally by running the gradle command `ktfmtFormat`

NudurupatiSurya commented:
Hello @shobhitagarwal1612,

Thank you for the feedback. I have made the changes you suggested. Please review and let me know if any further modifications are needed.

Regards,
Surya

CC: @gino-m

shobhitagarwal1612 commented:
/gcbrun

shobhitagarwal1612 commented:
The remote build is still failing. Can you please run `checkCode` locally and fix whatever is failing?

NudurupatiSurya commented:
> The remote build is still failing. Can you please run `checkCode` locally and fix whatever is failing?

I ran `checkCode` locally and found an unused resource file and some styles. I have removed these, and the build is now successful for `checkCode`. Do let me know if any further changes are needed. 😊

shobhitagarwal1612 commented:
/gcbrun

shobhitagarwal1612 commented:
@gino-m PTAL

gino-m commented:
/gcbrun
